### ISSUE TITLE
fix: a derived beacon wouldn't subscribe to writable beacon.

### ISSUE THREAD

**Issue by yehorh** [jinyus/dart_beacon#110]
**Description**
Strange behavior, what am I doing wrong?

**Steps To Reproduce**

Minimal reproducible version:
```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:state_beacon/state_beacon.dart';

typedef Themes = ({ThemeData lightTheme});

class Customization {
  const Customization({
    this.themes = const {},
  });

  final Map<String, Themes> themes;
}

class PrefController {
  PrefController({
    required this.defaultThemeName,
    required this.appCustomization,
  });

  final String? defaultThemeName;
  final Customization appCustomization;

  late final selectedTheme = Beacon.writable<String>(defaultThemeName ?? 'Default', name: 'selectedTheme');

  late final lightTheme = Beacon.derived(
    () {
      final selectedThemeName = selectedTheme.value;
      return appCustomization.themes[selectedThemeName]?.lightTheme;
    },
    name: 'lightTheme',
  );
}

const color1 = Color(0xFF000000);
const color2 = Color(0xFFFF0000);
const colorFallback = Color(0xFFFFFFFF);

const myWidgetKey = Key('myWidgetKey');

void main() {
  testWidgets('Should change theme with empty', (widgetTester) async {
    BeaconObserver.instance = LoggingObserver();
    final PrefController controller = PrefController(
      defaultThemeName: null,
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );

    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), 'Default');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, colorFallback);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });

  testWidgets('Should change theme with invalid predefined value', (widgetTester) async {
    BeaconObserver.instance = LoggingObserver();
    final PrefController controller = PrefController(
      defaultThemeName: 'value not exists',
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );

    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), 'value not exists');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, colorFallback);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });

  testWidgets('Should change theme with predefined value', (widgetTester) async {
    final PrefController controller = PrefController(
      defaultThemeName: '1',
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );

    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), '1');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });
}

class MyApp extends StatelessWidget {
  const MyApp({
    required this.controller,
    super.key,
  });

  final PrefController controller;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: controller.lightTheme.watch(context) ?? ThemeData.light().copyWith(primaryColor: colorFallback),
      home: Builder(
        builder: (context) {
          return Scaffold(
            body: ColoredBox(
              key: myWidgetKey,
              color: Theme.of(context).primaryColor,
            ),
          );
        },
      ),
    );
  }
}
```

**Comment by yehorh** [comment]
If I rewrite it due to effects, then it will all work.

But I don't understand why in the last test case it doesn't want to work.
It seems the widget doesn't want to observe the derived beacon.

<details>
  <summary>Example `effect implementation`</summary>
```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:state_beacon/state_beacon.dart';

typedef Themes = ({ThemeData lightTheme});

class Customization {
  const Customization({
    this.themes = const {},
  });

  final Map<String, Themes> themes;
}

class PrefController {
  PrefController({
    required this.defaultThemeName,
    required this.appCustomization,
  });

  void init() {
    Beacon.effect(() {
      final selectedThemeName = selectedTheme.value;
      lightTheme.value = appCustomization.themes[selectedThemeName]?.lightTheme;
    });
  }

  final String? defaultThemeName;
  final Customization appCustomization;

  late final selectedTheme = Beacon.writable<String>(defaultThemeName ?? 'Default', name: 'selectedTheme');

  late final lightTheme = Beacon.writable<ThemeData?>(null, name: 'lightTheme');
}

const color1 = Color(0xFF000000);
const color2 = Color(0xFFFF0000);
const colorFallback = Color(0xFFFFFFFF);

const myWidgetKey = Key('myWidgetKey');

void main() {
  testWidgets('Should change theme with empty', (widgetTester) async {
    final PrefController controller = PrefController(
      defaultThemeName: null,
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );
    controller.init();

    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), 'Default');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, colorFallback);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });

  testWidgets('Should change theme with invalid predefined value', (widgetTester) async {
    BeaconObserver.instance = LoggingObserver();
    final PrefController controller = PrefController(
      defaultThemeName: 'value not exists',
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );

    controller.init();
    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), 'value not exists');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, colorFallback);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });

  testWidgets('Should change theme with predefined value', (widgetTester) async {
    BeaconObserver.instance = LoggingObserver();
    final PrefController controller = PrefController(
      defaultThemeName: '1',
      appCustomization: Customization(
        themes: {
          '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
          '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
        },
      ),
    );

    controller.init();
    await widgetTester.pumpWidget(MyApp(controller: controller));
    expect(controller.selectedTheme.peek(), '1');
    var byKey = find.byKey(myWidgetKey);
    expect(byKey.first, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);

    controller.selectedTheme.value = '2';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '2');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color2);

    controller.selectedTheme.value = '1';
    await widgetTester.pumpAndSettle();
    expect(controller.selectedTheme.peek(), '1');
    byKey = find.byKey(myWidgetKey);
    expect(byKey, findsOneWidget);
    expect(widgetTester.widget<ColoredBox>(byKey).color, color1);
  });
}

class MyApp extends StatelessWidget {
  const MyApp({
    required this.controller,
    super.key,
  });

  final PrefController controller;

  @override
  Widget build(BuildContext context) {
    final themeData = controller.lightTheme.watch(context);
    return MaterialApp(
      theme: themeData ?? ThemeData.light().copyWith(primaryColor: colorFallback),
      home: Builder(
        builder: (context) {
          return Scaffold(
            body: ColoredBox(
              key: myWidgetKey,
              color: Theme.of(context).primaryColor,
            ),
          );
        },
      ),
    );
  }
}
```
</details>

**Comment by jinyus** [comment]
This should work, looking into it...

**Comment by jinyus** [comment]
I isolated the beacons and it works...Trying to figure out why the derived isn't picking up the update. 

 this is basically what it's doing:
```dart
test('derived should work', () async {
    BeaconObserver.useLogging();
    final appCustomization = Customization(
      themes: {
        '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
        '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
      },
    );

    late final selectedTheme = Beacon.writable<String>(
      'Default',
      name: 'selectedTheme',
    );

    late final lightTheme = Beacon.derived(
      () {
        final selectedThemeName = selectedTheme.value;
        return appCustomization.themes[selectedThemeName]?.lightTheme;
      },
      name: 'lightTheme',
    );

    var currentTheme = lightTheme.peek();

    lightTheme.subscribe(
      (v) => currentTheme = v,
      startNow: false,
    );

    expect(currentTheme, isNull);
    selectedTheme.value = '1';

    await Future.delayed(Duration(milliseconds: 100));

    expect(currentTheme!.primaryColor, color1);

    selectedTheme.value = '2';

    await Future.delayed(Duration(milliseconds: 100));

    expect(currentTheme!.primaryColor, color2);
  });
```

**Comment by yehorh** [comment]
> I isolated the beacons and it works...Trying to figure out why the derived isn't picking up the update when used in the controller.
> 
> this is basically what it's doing:
> 
> ```dart
> test('derived should work', () async {
>     BeaconObserver.useLogging();
>     final appCustomization = Customization(
>       themes: {
>         '1': (lightTheme: ThemeData.light().copyWith(primaryColor: color1)),
>         '2': (lightTheme: ThemeData.light().copyWith(primaryColor: color2)),
>       },
>     );
> 
>     late final selectedTheme = Beacon.writable<String>(
>       'Default',
>       name: 'selectedTheme',
>     );
> 
>     late final lightTheme = Beacon.derived(
>       () {
>         final selectedThemeName = selectedTheme.value;
>         return appCustomization.themes[selectedThemeName]?.lightTheme;
>       },
>       name: 'lightTheme',
>     );
> 
>     var currentTheme = lightTheme.peek();
> 
>     lightTheme.subscribe(
>       (v) => currentTheme = v,
>       startNow: false,
>     );
> 
>     expect(currentTheme, isNull);
>     selectedTheme.value = '1';
> 
>     await Future.delayed(Duration(milliseconds: 100));
> 
>     expect(currentTheme!.primaryColor, color1);
> 
>     selectedTheme.value = '2';
> 
>     await Future.delayed(Duration(milliseconds: 100));
> 
>     expect(currentTheme!.primaryColor, color2);
>   });
> ```

Yep. I have tried isolate it.
But could reproduce only with `watch` extension.

**Comment by jinyus** [comment]
Fixed in v0.45.2.

It was a weird edge case with subscriptions.
